### ISSUE TITLE
Error when implementing optimal representative sampling

### ISSUE THREAD

**Issue by sbarratt** [cvxgrp/a2dr#2]
The problem looks like

minimize I_{fdes}(x_1) + neg_entr(x_2) + I_{\ones^Tx=1}(x_3)
subject to x_1 = Fx_2, x_2 = x_3.

The code is:
```
import numpy as np
import pandas as pd
from a2dr import a2dr
from a2dr.proximal import *
from scipy import sparse

np.random.seed(5)
n = 100
sex = np.random.choice([0., 1.], p=[.4, .6], size=n)
height = np.random.normal(5, 1, size=n)

df = pd.DataFrame({
    "sex": sex,
    "height": height
})

F = np.array(df).T
m, n = F.shape
fdes = np.array([.5, 5.3])

def prox_f1(v, t=1):
    return fdes
def prox_sum_to_one(v, t=1):
    return v + (1 - v.sum()) * np.ones(v.size) / v.size**2

A_1 = sparse.vstack([-sparse.eye(2), sparse.csc_matrix((100, 2))]).tocsc()
A_2 = sparse.vstack([F, sparse.eye(100)]).tocsc()
A_3 = sparse.vstack([sparse.csc_matrix((2, 100)), -sparse.eye(100)]).tocsc()

a2dr_result = a2dr([prox_f1, prox_neg_entr, prox_sum_to_one], [A_1, A_2, A_3], b=np.zeros(102))
```

The output is:
```
----------------------------------------------------------------------
a2dr v0.2.3.post2 - Prox-Affine Distributed Convex Optimization Solver
                       (c) Anqi Fu, Junzi Zhang
                      Stanford University   2019
----------------------------------------------------------------------
### Preconditioning starts ... ###
### Preconditioning finished.  ###
max_iter = 1000, t_init (after preconditioning) = 0.00
eps_abs = 1.00e-06, eps_rel = 1.00e-08, precond = True
ada_reg = True, anderson = True, m_accel = 10
lam_accel = 1.00e-08, aa_method = lstsq, D_safe = 1.00e+06
eps_safe = 1.00e-06, M_safe = 10
variables n = 202, constraints m = 102
nnz(A) = 360
Setup time: 9.41e-03
----------------------------------------------------
 iter | total res | primal res | dual res | time (s)
----------------------------------------------------
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
<ipython-input-2-85b176437a88> in <module>
     22 A_3 = sparse.vstack([sparse.csc_matrix((2, 100)), -sparse.eye(100)]).tocsc()
     23 
---> 24 a2dr_result = a2dr([prox_f1, prox_neg_entr, prox_sum_to_one], [A_1, A_2, A_3], b=np.zeros(102))

~/miniconda3/lib/python3.7/site-packages/a2dr/solver.py in a2dr(p_list, A_list, b, v_init, n_list, *args, **kwargs)
    288         v_half = np.concatenate(v_halves, axis=0)
    289         sys.stdout = open(os.devnull, 'w')
--> 290         dk = sp.linalg.lsqr(A, A.dot(v_half) - b, atol=1e-10, btol=1e-10, x0=dk)[0]
    291         sys.stdout.close()
    292         sys.stdout = sys_stdout_origin

~/miniconda3/lib/python3.7/site-packages/scipy/sparse/linalg/isolve/lsqr.py in lsqr(A, b, damp, atol, btol, conlim, iter_lim, show, calc_var, x0)
    431         # This alters the diagonal (rhobar) of the lower-bidiagonal matrix.
    432         rhobar1 = sqrt(rhobar**2 + damp**2)
--> 433         cs1 = rhobar / rhobar1
    434         sn1 = damp / rhobar1
    435         psi = sn1 * phibar

ZeroDivisionError: float division by zero

     0| 2.23e+01    8.57e-03     2.23e+01   2.40e-02
```

Am I doing something wrong here?

**Comment by junziz** [comment]
Thanks a lot for catching this! There are mainly two issues, and both of them are specific to the negative entropy/KL proximal operators, and more specifically the lambertw and exponential functions inside them.

1. In the proximal operator implementation of prox_neg_entr and prox_kl, we didn't include np.real on top of the lambertw function, and lsqr throws an error for complex inputs (even if the imaginary part is 0). This has now been resolved in the current update by adding np.real in the two proximal operators.
2. After resolving the first issue above, there is still an issue of overflow in the exp function inside the lambertw function. And this is due to our preconditioning, which resets the proximal penalty parameter to 1e-3 or so, and inside \exp, the term v/t scales up to the order of 1e3 and hence blows up. I tried to turn off the preconditioning by setting precond=False, and then it works and converges in 0.2 seconds/36 iterations. 

In short, after reinstalling the updated code (or simply redefine you own prox_neg_entr function by adding np.real as mentioned above, instead of importing it from the a2dr proximal library), you can run your code with everything the same apart from the last line changed to 
```
a2dr_result = a2dr([prox_f1, prox_neg_entr, prox_sum_to_one], [A_1, A_2, A_3], b=np.zeros(102), precond=False)
```
 and the problem should go away. 

In general, preconditioning does not help in all cases and so it might be a good practice to avoid preconditioning/resetting t_init when seeing numerical overflow/errors (or other similar issues).

Please feel free to let us know if there are any additional problems! Thanks again!

**Comment by sbarratt** [comment]
OK, got it working! Thanks for such a prompt response.

```python
import numpy as np
import pandas as pd
from a2dr import a2dr
from a2dr.proximal import *
from scipy import sparse
import cvxpy as cp
import time
import rsw

np.random.seed(5)
n = 100
sex = np.random.choice([0., 1.], p=[.4, .6], size=n)
height = np.random.normal(5, 1, size=n)

df = pd.DataFrame({
    "sex": sex,
    "height": height
})

F = np.array(df).T
m, n = F.shape
fdes = np.array([.5, 5.3])

# Solve using A2DR
def prox_f1(v, t=1):
    return fdes
def prox_sum_to_one(v, t=1):
    return v + (1 - v.sum()) * np.ones(v.size) / v.size

A_1 = sparse.vstack([-sparse.eye(2), sparse.csc_matrix((100, 2))]).tocsc()
A_2 = sparse.vstack([F, sparse.eye(100)]).tocsc()
A_3 = sparse.vstack([sparse.csc_matrix((2, 100)), -sparse.eye(100)]).tocsc()

tic = time.time()
a2dr_result = a2dr([prox_f1, prox_neg_entr, prox_sum_to_one], [A_1, A_2, A_3], b=np.zeros(102), precond=False, verbose=False)
toc = time.time()
time_a2dr = toc - tic
w_a2dr = a2dr_result['x_vals'][1]

# Solve using CVXPY
w = cp.Variable(100)
prob = cp.Problem(cp.Minimize(-cp.sum(cp.entr(w))), [fdes == F @ w, w >= 0, cp.sum(w) == 1])
tic = time.time()
prob.solve(solver=cp.SCS)
toc = time.time()
time_cvxpy = toc - tic
w_cvxpy = w.value

# Solve using RSW
losses = [rsw.EqualityLoss(.5),
          rsw.EqualityLoss(5.3)]
regularizer = rsw.EntropyRegularizer()
tic = time.time()
w, out, sol = rsw.rsw(F.T, None, losses, regularizer, 1., rho=300, verbose=False, eps_abs=1e-8, eps_rel=1e-8)
toc = time.time()
w_rsw = w
time_rsw = toc-tic

# Test all close
np.testing.assert_allclose(w_rsw, w_cvxpy, atol=1e-5)
np.testing.assert_allclose(w_a2dr, w_cvxpy, atol=1e-5)
np.testing.assert_allclose(w_a2dr, w_rsw, atol=1e-5)

# Print times
print("Time rsw %.3f" % time_rsw)
print("Time a2dr %.3f" % time_a2dr)
print("Time cvxpy %.3f" % time_cvxpy)
```

For me, this outputs
```
Iteration: 0
Iteration: 10
Iteration: 20
Iteration: 30
Iteration: 40
Time rsw 0.069
Time a2dr 0.113
Time cvxpy 0.085
```
It seems that a2dr is printing even though I sent in verbose=False. You might want to look into this.

**Comment by junziz** [comment]
My pleasure! And thanks for spotting the verbose issue. Just fixed it in the newest update. However, the printout I saw before fixing the issue was only the last iterate's information of iteration number and primal-dual residuals, instead of "Iteration: k" as you showed above. Printouts of the form "Iteration: k" should have been only seen in the outdated versions of a2dr before it goes to PyPi. Please let me know if the printout issue is resolved after the current update. If not, would you mind checking the version of your a2dr, and in particular, if you are running a legacy version of a2dr instead of v0.2.3.post2? Thanks!

Also it seems that the runtimes are pretty dependent on the systems. The typical runtime on my personal computer is a2dr = 0.18s and cvxpy = 0.59s (on average), while it is 0.31s and 0.34s on the server. (I commented out the part of rsw for now.) For larger instances like n = 1000, on my personal computer a2dr takes 0.66s and cvxpy takes 2.59s, while on the server a2dr takes 2.9s and cvxpy takes 17.3s.
